Add support for isolator watch callback

komuta opened the pull request:
This PR adds support of the `watch` callback for the CommandIsolator.
This callback differs from `prepare` and `cleanup` in that it is supposed to be long running, and return only if the container has reached a certain limit and is to be killed. It can also returns with an abandoned future, in which case the container will not be killed (with the drawback that the `watch` callback will never be called again during the lifecycle of the container).

This implementation is experimental, so not adding the documentation to README for now.

komuta committed:
Add support for isolator watch callback

komuta committed:
Add missing files

kamaradclimber commented on src/CommandIsolator.cpp:
```diff
@@ -82,6 +90,40 @@ process::Future<Option<ContainerLaunchInfo>> CommandIsolatorProcess::prepare(
+                 << containerLimitation.error();
+    return process::Future<ContainerLimitation>();
+  }
+  return containerLimitation.get();
```
are you sure it matches the type Future<containerlimitation> ? it seems it returns a raw ContainerLimitation (I'm not expert on those types)

clems4ever commented on tests/scripts/watch_incorrect_protobuf.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+echo '"does not match ContainerLimitation"' >$2
```
This is supposed to test what happens if the JSON is not matching the model (JSON is valid but one key is incorrect for instance) but in this case you're doing the same test as the malformed one beneath (where you are testing the output is not a valid JSON).

komuta replied:
I'm not sure to understand: my `watch_malformed.sh` fixture outputs malformed JSON.
```echo 'MALFORMED' >$2```

The `watch_incorrect_protobuf.sh` mentioned above outputs correct JSON but not matching the `ContainerLimitation` protobuf model.
So I think I test both cases, even though they go through the exact same code path (inside `watch()`).

On the other side, there are still a few code paths not tested, I'll fix that.

clems4ever replied:
the malformed test is correct, the incorrect protobuf should be a valid json with wrong model.

komuta replied:
This is the case. [A string is a valid JSON](https://www.json.org/).

clems4ever replied:
Got it. I hope they follow the standard :).

clems4ever reviewed (changes requested):
Can you clarify what does it mean that the watch method is different in the fact that it is "long-running". I don't see any loop or anything making it long running, it returns as other calls. Can you clarify?

komuta commented:
> Can you clarify what does it mean that the watch method is different in the fact that it is "long-running". I don't see any loop or anything making it long running, it returns as other calls. Can you clarify?

There is indeed no loop: the `watch` call returns a `Future<ContainerLimitation>`, is called once by Mesos containerizer which will trigger a call to limit (kill and cleanup) the container when the future is confirmed and ready. This means that `watch` is intended to launch a long running, watching process (in the meaning of libprocess) to check frequently the status of the container.

Mwea commented on src/CommandIsolator.cpp:
```diff
@@ -82,6 +90,40 @@ process::Future<Option<ContainerLaunchInfo>> CommandIsolatorProcess::prepare(
   return containerLaunchInfo.get();
 }
 
+process::Future<ContainerLimitation> CommandIsolatorProcess::watch(
```
Why don't you use an optional as the prepare command does ?

komuta replied:
`None()` would be considered as something, so it won't match [the default constructor](https://github.com/criteo-forks/mesos/blob/criteo/1.7.0/3rdparty/libprocess/include/process/future.hpp#L1063-L1068) which is the only one setting the `Future` as *abandoned*.
The *abandoned* state is important because it is the only one which won't trigger a limitation from the containerizer (neither killing the container nor reporting a failure).

komuta committed:
Add a watch test handling empty command output

komuta committed:
Nasty hack to allow to ignore timeout

This is necessary for the time being, to allow watch command to run
forever.
It can be removed once the new implementation of watch callback (polling
the watch command regularly) will be done.

clems4ever commented on src/CommandRunner.cpp:
```diff
@@ -249,23 +249,28 @@ Try<Nothing> runCommandWithTimeout(const std::string& command,
+  // Hack to support command without timeout
+  // TODO: to remove once the CommandIsolator::watch implementation has been
+  // changed
+  if (timeoutInSeconds > 0) {
```
Why do you need to skip the SIGTERM?

clems4ever replied:
would not it be easier to just set the timeout to 1 and avoid complexifying this part?

komuta replied:
I oversaw the termination handling part.
We can't set it to 1, the command has to be running for as long as the container is here. I'll think about something, or discard this part, knowing that an implementation change is on its way afterwards.

komuta committed:
Revert "Add a watch test handling empty command output"

This reverts commit 29007ac824ace12b5b748011c6795a4e79ddf7a5.

komuta committed:
Revert "Revert "Add a watch test handling empty command output""

This reverts commit 85f3bef08fd2131b805a7181b6e8c7c085e55f1f.

komuta committed:
Revert "Nasty hack to allow to ignore timeout"

This reverts commit 69a3c62a1bf4a4d338aba4dc5411a5756f12a8c7.
